pkg/deploy/controller/deployment/controller_test.go

```diff
@@ -651,6 +651,81 @@ func TestHandle_deployerPodDisappeared(t *testing.T) {
 	}
 }
 
+func expectMapContains(t *testing.T, exists, expected map[string]string, what string) {
+	if expected == nil {
+		return
+	}
+	for k, v := range expected {
+		value, ok := exists[k]
+		if ok && value != v {
+			t.Errorf("expected %s[%s]=%s, got %s", what, k, v, value)
+		} else if !ok {
+			t.Errorf("expected %s %s: not present", what, k)
+		}
+	}
+}
+
+func TestDeployerCustomLabelsAndAnnotations(t *testing.T) {
+	var createdPod *kapi.Pod
+
+	controller := &DeploymentController{
+		decodeConfig: func(deployment *kapi.ReplicationController) (*deployapi.DeploymentConfig, error) {
+			return deployutil.DecodeDeploymentConfig(deployment, api.Codec)
+		},
+		deploymentClient: &deploymentClientImpl{
+			updateDeploymentFunc: func(namespace string, deployment *kapi.ReplicationController) (*kapi.ReplicationController, error) {
+				return deployment, nil
+			},
+		},
+		podClient: &podClientImpl{
+			createPodFunc: func(namespace string, pod *kapi.Pod) (*kapi.Pod, error) {
+				createdPod = pod
+				return pod, nil
+			},
+		},
+		makeContainer: func(strategy *deployapi.DeploymentStrategy) (*kapi.Container, error) {
+			return okContainer(), nil
+		},
+		recorder: &record.FakeRecorder{},
+	}
+
+	testCases := []struct {
+		strategy    deployapi.DeploymentStrategy
+		labels      map[string]string
+		annotations map[string]string
+	}{
+		{deploytest.OkStrategy(), map[string]string{"label1": "value1"}, map[string]string{"annotation1": "value1"}},
+		{deploytest.OkCustomStrategy(), map[string]string{"label2": "value2", "label3": "value3"}, nil},
+		{deploytest.OkCustomStrategy(), nil, map[string]string{"annotation3": "value3"}},
+		{deploytest.OkStrategy(), map[string]string{deployapi.DeployerPodForDeploymentLabel: "ignored"}, nil},
+	}
+
+	for c, test := range testCases {
+		config := deploytest.OkDeploymentConfig(1)
+		config.Template.Strategy = test.strategy
+		config.Template.Strategy.Labels = test.labels
+		config.Template.Strategy.Annotations = test.annotations
+
+		deployment, _ := deployutil.MakeDeployment(config, kapi.Codec)
+		deployment.Annotations[deployapi.DeploymentStatusAnnotation] = string(deployapi.DeploymentStatusNew)
+
+		err := controller.Handle(deployment)
+		if err != nil {
+			t.Fatal(err)
+		}
+		nameLabel, ok := createdPod.Labels[deployapi.DeployerPodForDeploymentLabel]
+		if ok && nameLabel != deployment.Name {
+			t.Errorf("label %s expected %s, got %s", deployapi.DeployerPodForDeploymentLabel, deployment.Name, nameLabel)
+		} else if !ok {
+			t.Errorf("label %s not present", deployapi.DeployerPodForDeploymentLabel)
+		}
+		if c != 3 {
```

Comment:
Maybe I misinterpreted why you are special casing scenario index 3. Whatever the reason, you shouldn't need to special case a specific index in the test case array. Otherwise when sombody inserts a new one in the middle later on, the assumption will break, likely surprising the author.

Just saying you should find a way to express this without having to hard code logic for a single test scenario.